### PR TITLE
Added annotation tool and fixed arrow [#178847796]

### PR DESCRIPTION
**Pull request by dougmartin**
The annotation tool is an optional tool that allows the user to enter text and place a connected arrow to the area to annotate.

This also fixes the arrow tool serialization which was broken in the upgrade of Fabric.js from v2 to v3.

### CODE REVIEW

File: app/scripts/tools/shape-tools/annotation-tool.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+function AnnotationTool(name, drawTool) {
+  ShapeTool.call(this, name, drawTool);
+
+  this.canvas.on(
+    'text:editing:exited',
+    function (opt) {
+      if (this.active) {
+        // This may be confusing, but if you take a look at FabricJS source, you will notice
+        // that text .selectable property is always set to true just before this event
+        // is emitted. Quite often is now what we want, especially when AnnotationTool is active.
```
Comment (kswenson) [inline app/scripts/tools/shape-tools/annotation-tool.js:19]:
```suggestion
        // is emitted. Quite often this is not what we want, especially when AnnotationTool is active.
```

File: app/scripts/tools/shape-tools/annotation-tool.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  // (drawing tool inside jQuery UI modal dialog).
+  // Note that it's exactly the same what FabricJS does in IText onMouseDown handler
+  // (at least in FabricJS v1.4.11).
+  if (text.hiddenTextarea && text.canvas) {
+    text.canvas.wrapperEl.appendChild(text.hiddenTextarea);
+    if (fabric.isTouchSupported) {
+      // Mobile devices seem to automatically zoom and scroll to input when it gets focus. Set a big font size to
+      // avoid zooming. Also, set correct 'top' value so page scrolls to the correct position. Note that this solution
+      // isn't perfect, as still page might scroll to the left (due to -1000px val). Unfortunately we can't keep
+      // hidden text input in the right place, as the input caret seems to ignore z-index and its always visible (iOS).
```
Comment (kswenson) [inline app/scripts/tools/shape-tools/annotation-tool.js:136]:
```suggestion
      // hidden text input in the right place, as the input caret seems to ignore z-index and it's always visible (iOS).
```

File: app/scripts/tools/shape-tools/annotation-tool.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    window.addEventListener('mousedown', handler, true);
+    window.addEventListener('touchstart', handler, true);
+  }
+  function cleanupHandlers() {
+    window.removeEventListener('mousedown', handler, true);
+    window.removeEventListener('touchstart', handler, true);
+  }
+  function handler(e) {
+    // By default when you click any element, active text should exit edit mode.
+    // However if clicked element (or his ancestor) has special class 'dt-keep-text-edit-mode',
```
Comment (kswenson) [inline app/scripts/tools/shape-tools/annotation-tool.js:167]:
```suggestion
    // However if clicked element (or its ancestor) has special class 'dt-keep-text-edit-mode',
```

**Review by kswenson — COMMENTED** [review]
Some impressive sleuthing here!

**Review by knowuh — APPROVED** [review]
Wow, this reminds me a bit of hacking JS Plumb.

Any way, looks good to me.

**Comment by dougmartin** [comment]
> Some impressive sleuthing here!

@kswenson I can't take any credit for the text - this is copied from the existing text tool. :)

**Comment by pjanik** [comment]
@dougmartin, this already got merged in, but when I was looking at this, I found two issues:
1. Any undo action breaks control points for existing annotation. So, it's no longer possible to move the arrow after using undo button.
2. When you select multiple annotations (or annotations + other objects) and move them around, control points move, but arrows don't. So, the tip of the arrow is out of sync with the control point. If you move the control point, it'll fix itself.

I don't know how important these issues are and how difficult they would be to fix. It seems to me that 1 is more important, while 2 is a glitch that users can workaround.

**Comment by dougmartin** [comment]
@pjanik Thanks - I will look at those issues when I'm done updating the UI today.

**Pull request merged**